sdks/python/apache_beam/runners/portability/local_job_service.py

```diff
@@ -95,7 +95,7 @@ def create_beam_job(self, preparation_id, job_name, pipeline, options):
 
   def start_grpc_server(self, port=0):
     self._server = grpc.server(UnboundedThreadPoolExecutor())
-    port = self._server.add_insecure_port('localhost:%d' % port)
+    port = self._server.add_insecure_port('[::]:%d' % port)
```

Comment:
Is this for a new use case? Originally, the embedded job server is intended for testing pipeline execution on the same machine or container. `localhost` should work fine for this case.

Reply:
A slightly new use case.  Instead of throwing away the job server for each invocation, we'd like to keep it around for multiple jobs, like we do with the Flink job server, then we can use our job-server-ui to view and manage jobs there.  

Additionally, we are sub-classing this class (`LocalJobServicer`) to make a new kind of job server that is able to proxy other job servers - a middle-man for reviewing a job before sending it on to Flink, Direct, Spark, etc. 

As a result we want to stand this service up along with our other long-running Beam services.

Reply:
Could this not be handled in the subclass? I think the notion of the `LocalJobServer` is not to listen on all interfaces. We could make the bind address configurable.

Reply:
> Could this not be handled in the subclass?

Yeah, let me look into the best design for this.  It'd be nice if we took this opportunity to make more than the hostname configurable (i.e. provide a way to further configure the server before starting).  I'll propose an alternative later today.

Reply:
I think it'd make sense for this to be parameterized, likely with localhost as a default.

Reply:
ok, I've been playing around with this, and the main impediment to keeping this design simple is that there are two separate hostnames required, one for opening the port for the server, and one which is delivered to the client for reconnecting to the staging service.

I think it'd be nice to prevent people from having to figure out all this stuff again, because it's pretty frustrating to get it right, so here's my best effort at a compromise between making this configurable and making it "just work".

```python
  def get_hostname(self):
    """Return the host name at which this server will be accessible.

    In particular, this is provided to the client as the
    artifact staging endpoint.
    """
    return 'localhost'

  def start_grpc_server(self, port=0):
    self._server = grpc.server(UnboundedThreadPoolExecutor())
    hostname = self.get_hostname()
    # either open this up to the world, or lock it down to localhost
    if os.environ.get('DOCKER_MAC_CONTAINER') == '1' or hostname != 'localhost':
      service_address = '[::]'
    else:
      service_address = 'localhost'
    port = self._server.add_insecure_port('%s:%d' % (service_address, port))
    beam_job_api_pb2_grpc.add_JobServiceServicer_to_server(self, self._server)
    beam_artifact_api_pb2_grpc.add_ArtifactStagingServiceServicer_to_server(
        self._artifact_service, self._server)
    self._artifact_staging_endpoint = endpoints_pb2.ApiServiceDescriptor(
        url='%s:%d' % (hostname, port))
    self._server.start()
    _LOGGER.info('Grpc server started at %s on port %d' % (hostname, port))
    return port
```

What do you think?  The other option is that I just copy all of `start_grpc_server` into my sub-class.  It's not the end of the world if that's the decision we come to.

Reply:
It looks ok, but one idea:

>there are two separate hostnames required, one for opening the port for the server, and one which is delivered to the client for reconnecting to the staging service.

Wouldn't it work to allow configuration of a `bind` address (e.g. `[::]`) and a `connect` address (e.g. `service_name`, what is currently returned in `get_hostname()`)? That way we would give the subclass full control over what it wants to do. Maybe return this as a tuple to avoid multiple overrides.

Reply:
Yeah, that's probably the most future-proof and versatile solution. 

What do you think about terminology for this?

1. `get_bind_address()`, `get_connect_address()`
2. `get_bind_address()`, `get_service_address()`
3. `get_bind_address()`, `get_hostname()`

Something else?

Reply:
```python
  def get_bind_address(self):
    """Return the address used to open the port on the gRPC server.
    """
    return self.get_connect_address()

  def get_connect_address(self):
    """Return the host name at which this server will be accessible.

    In particular, this is provided to the client upon connection as the
    artifact staging endpoint.
    """
    return 'localhost'

  def start_grpc_server(self, port=0):
    self._server = grpc.server(UnboundedThreadPoolExecutor())
    port = self._server.add_insecure_port('%s:%d' % (self.get_bind_address(), port))
    beam_job_api_pb2_grpc.add_JobServiceServicer_to_server(self, self._server)
    beam_artifact_api_pb2_grpc.add_ArtifactStagingServiceServicer_to_server(
        self._artifact_service, self._server)
    hostname = self.get_connect_address()
    self._artifact_staging_endpoint = endpoints_pb2.ApiServiceDescriptor(
        url='%s:%d' % (hostname, port))
    self._server.start()
    _LOGGER.info('Grpc server started at %s on port %d' % (hostname, port))
    return port
```

Reply:
Yes, that looks good!

Reply:
Let me know when you update the PR.

Reply:
@mxm updated